src/librustc/middle/const_eval.rs

```diff
@@ -361,10 +538,13 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
               ast::BiBitOr => a | b,
               ast::BiEq => a == b,
               ast::BiNe => a != b,
-              _ => return Err("can't do this op on bools".to_string())
+              _ => return Err(invalid_op_for_bools(e, op.node)),
              }))
           }
-          _ => Err("bad operands for binary".to_string())
+          (err @ Err(..), _) |
+          (_, err @ Err(..)) => err,
+
+          _ => Err(misc_binary_op(e)),
```

Comment:
uh, you are referring to the catch-all at the end that returns `Err(misc_binary_op(e))`, right?  I had assumed that this is trying to catch e.g. type-mismatches in the two results ... I do not understand what alternative you are suggesting to that here: the two sides could both hand back `Ok`'s.